### PR TITLE
Simulator: cleanup in preparation for metadrive

### PR DESCRIPTION
**Pull request by jnewb1**
With the goal of adding metadrive as another simulator. Mostly taken from https://github.com/commaai/openpilot/pull/27373

- bridge -> run_bridge, with --simulator argument that can switch between multiple possible simulators
- bridge/common.py for everything thats generic, and a SimulatorBridge base class
- bridge/carla.py for everything specific to carla, and CarlaBridge class
- VehicleState -> SimulatorState, which contains everything we need to give OP feedback and prevents passing a bunch of arguments to a bunch of functions
- use openpilot's engaged variable and send manual gas/brake values back to OP to disengage
- create simulated_car and simulated_sensors for sending all the feedback to OP

### CODE REVIEW

File: selfdrive/controls/controlsd.py [hunk 1]
```diff
@@ -339,7 +339,7 @@ def update_events(self, CS):
           self.events.add(EventName.cameraMalfunction)
         elif not self.sm.all_freq_ok(self.camera_packets):
           self.events.add(EventName.cameraFrameRate)
-    if not REPLAY and self.rk.lagging:
+    if not REPLAY and not SIMULATION and self.rk.lagging:
```
Comment (adeebshihadeh) [inline selfdrive/controls/controlsd.py:342]: this shouldn't be ignored. openpilot doesn't need much to run realtime

**Comment by jnewb1** [comment]
@adeebshihadeh this is all the cleanup stuff for carla, which will allow metadrive to be a drop in replacement

File: selfdrive/controls/controlsd.py (hunk 1 above)
Comment (jnewb1) [reply selfdrive/controls/controlsd.py]: done

File: tools/sim/bridge/carla.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+  def __init__(self, arguments):
+    super().__init__(arguments)
+    self.host = arguments.host
+    self.port = arguments.port
+    self.town = arguments.town
+    self.num_selected_spawn_point = arguments.num_selected_spawn_point
+
+  def spawn_world(self):
+    import carla  # pylint: disable=import-error
```
Comment (adeebshihadeh) [inline tools/sim/bridge/carla.py:126]: the pylint suppression needed?

File: tools/sim/bridge/carla.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    def connect_carla_client(host: str, port: int):
+      client = carla.Client(host, port)
+      client.set_timeout(5)
+      return client
+
+    client = connect_carla_client(self.host, self.port)
+    world = client.load_world(self.town)
+
+    settings = world.get_settings()
+    #settings.synchronous_mode = True  # Enables synchronous mode
```
Comment (adeebshihadeh) [inline tools/sim/bridge/carla.py:136]: is this dead?

File: tools/sim/bridge/carla.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    physics_control = vehicle.get_physics_control()
+    physics_control.mass = 2326
+    physics_control.torque_curve = [[20.0, 500.0], [5000.0, 500.0]]
+    physics_control.gear_switch_time = 0.0
+    vehicle.apply_physics_control(physics_control)
+
+    return CarlaWorld(world, vehicle, dual_camera=self.dual_camera)
+
+  def close(self):
+    super().close()
```
Comment (adeebshihadeh) [inline tools/sim/bridge/carla.py:171]: just don't override it?

File: tools/sim/bridge/carla.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+  def __init__(self, arguments):
+    super().__init__(arguments)
+    self.host = arguments.host
+    self.port = arguments.port
+    self.town = arguments.town
+    self.num_selected_spawn_point = arguments.num_selected_spawn_point
+
+  def spawn_world(self):
+    import carla  # pylint: disable=import-error
+    def connect_carla_client(host: str, port: int):
```
Comment (adeebshihadeh) [inline tools/sim/bridge/carla.py:127]: why the function?

File: tools/sim/bridge/carla.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
+import numpy as np
+from openpilot.common.params import Params
```
Comment (adeebshihadeh) [inline tools/sim/bridge/carla.py:2]: breakup the openpilot imports with a new line

File: tools/sim/bridge/carla.py (hunk 2 above)
Comment (jnewb1) [reply tools/sim/bridge/carla.py]: removed

File: tools/sim/bridge/carla.py (hunk 3 above)
Comment (jnewb1) [reply tools/sim/bridge/carla.py]: runs much better without it, removed the commented line

File: tools/sim/bridge/carla.py (hunk 4 above)
Comment (jnewb1) [reply tools/sim/bridge/carla.py]: meant to call world.close, done

File: tools/sim/bridge/carla.py (hunk 5 above)
Comment (jnewb1) [reply tools/sim/bridge/carla.py]: removed

File: tools/sim/bridge/carla.py (hunk 6 above)
Comment (jnewb1) [reply tools/sim/bridge/carla.py:3]: done

**Review by adeebshihadeh — APPROVED** [review]

**Pull request merged**